File: src/data-structures/ts/Queue.ts
type Node<T> = {
  value: T,
  next?: Node<T>
}

export default class Queue<T> {
  public length: number;
  private head?: Node<T>;
  private tail?: Node<T>;

  constructor() {
    this.head = this.tail = undefined;
    this.length = 0;
  }

  enqueue(item: T): void {
    this.length++;
    const node: Node<T> = {
      value: item,
      next: undefined
    }
    if (this.length === 1) {
      this.tail = this.head = node;
    }

    const tail = this.tail as Node<T>;
    tail.next = node;
    this.tail = node;
  }

  deque(): T | undefined {
    if (!this.head) return;

    this.length--;

    const head = this.head as Node<T>;
    this.head = head.next;

    if (this.length === 0) {
      this.tail = this.head = undefined;
      return head.value;
    }

    head.next = undefined;
    
    return head.value;
  }

  peek(): T | undefined {
    return this.head?.value;
  }
}

File: src/data-structures/java/Queue.java
public class Node<T> {
  T val;
  Node<T> next;
}

public class Queue<T> {

  int length;
  Node<T> tail;
  Node<T> head;

  public Queue() {
    this.head = head;
    this.tail = tail;
    this.length = 0;
  }

  public void Enqueue(T item) {
    if (item == null) return;
    this.length++;
    Node<T> newNode = new Node<T>();
    newNode.val = item;
    if (this.length == 1) {
      this.tail = this.head = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode; 
    }
  }

  public T Deque() {
    if (this.length == 0) return;
    this.length--;

    Node<T> prevHead = this.head;
    prevHead.next = null;

    this.head = this.head.next;
    if (this.length == 0) {
      this.tail = null;
    }
    return prevHead.val;
  }

  // peek
  public T Peek() {
    if(this.length == 0) return null;
  }

}